blockchain/contracts/WavGame.sol

```diff
@@ -0,0 +1,270 @@
+
+    
+    event Collected(address caller, address to, uint256 amount, uint256 refunded, uint256 totalMinted);
+    event LeveledUp(address caller, address collector, uint256 nextLevel, uint256 totalMinted);
+    event SpecialMint(address caller, address collector, uint256 id, uint256 amount);
+    event PaymentForwarded(address to, uint256 amount);
+    event GameSet(address owner, uint256 level);
+    event LevelSet(address owner, uint256 level);
+
+	// Create all initial artists
+	// create and map the 6 artist level / island to each artist, and define the Levels specific details per artist
+    constructor(IWavContract _wavContract, address _trustedForwarder,uint256 _entryFee, address payable _revenueContract) {
+        wavContract = _wavContract;
+        entryFee = _entryFee;
+        revenueContract = _revenueContract;
+        _setTrustedForwarder(_trustedForwarder);
+    }
+
+    // constructor(IWavContract _wavContract, uint256 _entryFee, address payable _revenueContract) {
+    //     wavContract = _wavContract;
+    //     entryFee = _entryFee;
+    //     revenueContract = _revenueContract;
+    // }
+
+
+    /// @notice This function mints level 1 NFTs to the recipient
+    /// @param _recipient The recipient of the minted NFT
+    /// @param _input ids and their corresponding quantities to mint
+    /// @dev This function does not support gasless, has to be called directly by user, since it requires payment in ETH and refund is possible
+    function collect(address _gameOwner, address _recipient, IDParam[] calldata _input) external override payable nonReentrant {
+        require(msg.value > 0, "PAYMENT_REQUIRED");
+        require(Helper.getLevelIndex(ENTRY_LEVEL) < wavGames[_gameOwner].length, "LEVEL_DOES_NOT_EXIST");
+
+        uint totalPayable = 0;
+        uint totalMinted = 0;
+        uint[] memory mintableIds = new uint[](_input.length); 
+        uint[] memory mintableAmountPerId = new uint[](_input.length);
+
+        // Create mintable IDs and valid quantity based on mintable set for base and msg.value
+        for (uint i; i < _input.length;) { 
+            require(mintableSet[_gameOwner][ENTRY_LEVEL].contains(_input[i].id), "NFTs_NOT_MINTABLE_FOR_LEVEL");
+            uint256 mintCost = _input[i].amount * entryFee;
+            if ((totalPayable + mintCost) > msg.value) { 
+                revert("INSUFFICIENT_MINT_AMOUNT");
+            }
+            mintableIds[i] = _input[i].id;
+            mintableAmountPerId[i] = _input[i].amount;
+            totalMinted +=_input[i].amount;
+            totalPayable += mintCost;
+            unchecked {++i;}
+        }
+        // Payment split
+        _forwardValue(totalPayable);
+        // Mint NFT to _recipient
+        wavContract.mintBatch(_recipient, mintableIds, mintableAmountPerId, bytes(" "));
+        _syncMint(_gameOwner, ENTRY_LEVEL, _recipient, totalMinted);
+        // Refund any balance
+        uint256 refund =  msg.value - totalPayable;
```

Comment:
Spending that much gas to get to this point only to revert because more than expected funds were sent in was my thought process for this. We can alternatively proceed without a refund, also, the refund is a one-off action, if it fails, no further action is taken. But I am curious as to why you think this might be risky.